### DISCUSSION TITLE
testing event loop to use io_uring instead of epoll

### DISCUSSION THREAD

**Discussion by kyoung-seo** [memcached/memcached#1190] [General]
Hello,

I have been trying to rigorously measure performance characteristic of io_uring based network multiplexing by porting various applications' event loop to use io_uring.

I have some preliminary result for memcached here [0].

It would be great if people who are familiar with memcached internals take a look at my experiment and see if the results look plausible with respect to their experience and intuition.

I used mutilate as a primary benchmark but I am not even sure if it is load generating scenario that is relevant. If there are other stress tests I could use I am all ears.

Regards,

[0] https://git.uwaterloo.ca/lseo/io_uring-experiments

**Comment by dormando** [comment]
Hey! Thanks for this (and reaching out). I did take a quickish look at it the other day.

So I did a similar experiment with a backend of the newer proxy code a few years ago, and found that under stress it wasn't really any faster than epoll; it had decent characteristics but didn't use less CPU. It sounded like features that were unavailable at the time (buffers/msg/etc) might help but it wasn't worth the extra code so I ripped it out. You can find it in the git history.

For performance testing I used to use `mc-crusher`, but the primary goal of that tool was to overload the system so I can find race conditions, stress locks, etc. It wasn't a good "load emulation tool". In the last year I've been using mcshredder: https://github.com/memcached/mcshredder - which has a flexible programming interface for load emulation. I've been building burn-in test suites on top of this. `mcshredder` is also my personal playground for io_uring; it's 100% io_uring and I was hoping to gain experience developing a new core loop there and then port that back to mc later.

I think though, as far as your experiments go, the primary issue is you've shut off extstore and TLS support. A bulk of deployments rely on one or the other. I've recently integrated OpenSSL with mcshredder on io_uring and it was _not_ great. It's extremely complicated to get OpenSSL out of the way enough to use all of io_uring's features, and so far as I can tell there're zero good public examples to go off of.

IE: without using your own BIO's OpenSSL's going to be mempcy'ing everything 3-4x. If you carefully use BIO's it seems like you can cut that down but it was difficult for me to do alone and nobody wants to pay for the work. It _seems_ like OpenSSL may need some patches to cut down excess memory copying even further.

Curious on your thoughts on this though. I've been flailing in the dark on io_uring and overall incredibly frustrated with it.

Edit: Sorry that was a bit rambly but my core point is it's hard to say if your results were strictly relevant because you're not using features that are critical for memcached.

**Reply by kyoung-seo** [reply]
Hello Dormando!

> So I did a similar experiment with a backend of the newer proxy code a
> few years ago, and found that under stress it wasn't really any faster
> than epoll; it had decent characteristics but didn't use less CPU. It
> sounded like features that were unavailable at the time
> (buffers/msg/etc) might help but it wasn't worth the extra code so I
> ripped it out. You can find it in the git history.

I see. Thank you for pointers. I will take a look.

> In the last year I've been using mcshredder:
> https://github.com/memcached/mcshredder - which has a flexible
> programming interface for load emulation. I've been building burn-in
> test suites on top of this. mcshredder is also my personal playground
> for io_uring; it's 100% io_uring and I was hoping to gain experience
> developing a new core loop there and then port that back to mc later.

Very nice! This looks much more flexible than mutilate. I will play with
it and see what happens. I am curious if you tried batching here or not as I
could not find any multishot related stub on the currently committed
code.

> I think though, as far as your experiments go, the primary issue is
> you've shut off extstore and TLS support. A bulk of deployments rely on
> one or the other. 

Right this was kind of insights I was looking for. I do not really have
any experience on how memcached is deployed in real world (I am only
using it because my advisor used memcached as a playground for his
network related experiments) so my initial experiment was more about
seeing if is it even worth doing serious research on io_uring as nobody
seemed to use it in production (outside of Meta) so I was not sure how
much was hype and how much was substantial.

Now I am half convinced that io_uring's infrastructure combined with batching
has some promising lead so I have been planning about my next move.

My university's clusters are old so extstore seems to be no go so I
should enable TLS, which brings me to the next elephant in the room.

> I've recently integrated OpenSSL with mcshredder on io_uring and it
> was not great. It's extremely complicated to get OpenSSL out of the
> way enough to use all of io_uring's features, and so far as I can tell
> there're zero good public examples to go off of.
> 
> IE: without using your own BIO's OpenSSL's going to be mempcy'ing
> everything 3-4x. If you carefully use BIO's it seems like you can cut
> that down but it was difficult for me to do alone and nobody wants to
> pay for the work. It seems like OpenSSL may need some patches to cut
> down excess memory copying even further.

Yeah this was the reason why I initially gave up making http proxy to
use io_uring and focus on memcached.

There is no consensus on how to deal with stream algorithms in io_uring
and well since nobody actually uses io_uring publicly, we don't know
what is the right idiom to use yet :)

I studied Christian's code [0] that I stumbled upon here [1] but I am
still not convinced what is the right way to do it.

> Curious on your thoughts on this though. I've been flailing in the dark
> on io_uring and overall incredibly frustrated with it.
> 
> Edit: Sorry that was a bit rambly but my core point is it's hard to say
> if your results were strictly relevant because you're not using features
> that are critical for memcached.

Yeah I think most people who are trying to use io_uring seriously are on
the same boat, which was partially the reason why I started this
experiment. To start some conversation with people to figure out where
we are at and where should we be going.

(It is a bit concerning that my code is still one of the only complete
example that demonstrates batching idioms in "real world" application
like memcached.)

I apologize for the fact that I don't really have any concrete answers
on hard problems you and I are facing but at least it is a good reality
check.

I guess I might just have to bite a bullet and see how far I can go with
TLS + BIO and then wee what Jens think. It would be very helpful if we
could see what Meta engineers are doing but their works are private so
everybody is going in blind...

[0] https://github.com/cmazakas/fiona
[1] https://github.com/axboe/liburing/discussions/1093

Edit: oops I thought BIO stood for "Batched IO" for some reason.

**Reply by dormando** [reply]
Yeah so overall I think we could've had _some_ gains by integrating io_uring a few years ago, but since TLS and extstore and etc it was just more work than it was worth. I think people hyped it up as like a 10xer thing but (especially a few years ago) for networking it certainly wasn't. I might try your branch on your test machine and see if I can replicate yours results but they seem a bit optimistic based on what I saw with my proxy code.

I pushed an `oldtls` branch up for mcshredder which has my flailing attempt at using OpenSSL BIO's. The main code tree is structured much better but I didn't have enough bookable hours to port the BIO stuff... let alone prove it did anything.

Someone put a PR converting extstore to io_uring a few years ago but it still used a side thread, so mostly it just moved the thread pool from userland to kernel land for those IO's... but it'd make a lot more sense if the core loops were all io_uring and the extstore datafile could be accessed directly from worker thread uring's instead of having to message pass... which is, again, a ton of work since we have to convert _everything_ else over.

My impression after fighting with it for three years is the general gains for a network app are in the 0-10% range if the application is already well optimized, and that level of improvement is only a big deal for companies with a footprint like meta. Of which there are few.

I haven't done the registered buffers stuff but my impressions is that's another maybe 5%... there's nothing io_uring can do that 10x or even 2x's your network code, unless I'm mistaken.

**Reply by dormando** [reply]
I may simply be too stupid to get multishot recv/registered buffers to work right, especially when mixed with BIO's. that's another problem.

If you're implementing TLS BIO stuff it'd probably be easier to start with mcshredder. multishot receive might still make sense (clients need to read responses, so it's the same problem but inverted) and it's already got TLS integrated... I've also iterated on the loop a few times (CQE handling is much better than it used to be, and it's not testing for NULL SQE's constantly anymore)

Edit: My long term plan is to make everything work in mcshredder, then also give mcshredder epoll/kqueue/poll/etc support and drop libevent. So then I can hopefully have a unified event structure, that's also smaller than what libevent has. Currently it would need to be both to support both uring/epoll/etc at once, which it has to since we're multiplatform software.

**Reply by kyoung-seo** [reply]
I tend to agree.

I see some scenarios where batched "completion based event loop" could be genuinely better than epoll like "readiness based event loop" [0] but it does not appear to be the worldbreaking magic wand that some people thought it was.

Then again, we are in catch-22 situation because more people need to seriously use io_uring for us to make scientific conclusions.

[0] https://github.com/golang/go/issues/65064#issuecomment-1889140389

**Reply by dormando** [reply]
Ironically I was looking at that same Go issue last week: io_uring might beat it there by a lot, but certainly not for memcached because we use one epoll context per worker thread already. Go might have similar solutions once they think it through.

**Comment by kyoung-seo** [comment]
This is unrelated but I thought I may as well ask you.

Initially I thought I was observing 50+% increase in throughput for the io_uring based event loop which looked way too good to be true.

It turned out that memcached ends up underclocking cpu when cpu frequency governor is schedutil not performance. Have you observed something like this before?

I know schedutil suffered some performance regressions before but I just assumed it got better recently. It was somewhat similar to issue described here[0].

[0] http://smalldatum.blogspot.com/2024/07/make-myrocks-2x-faster-by-changing-cpu.html

**Reply by dormando** [reply]
Memcached isn't doing anything with the clock frequency but yes that can happen. When testing I always set the governor to performance and I monitor the CPU frequencies during the first few test runs to ensure it's not downclocking or flipping states around frequently.

Usually you see it with latency being high at low request rates, then it suddenly drops as load increases and the clock starts to stabilize high.

I don't know any reason why the symptom you saw would happen more with io_uring.

**Reply by kyoung-seo** [reply]
I meant that the default memcached version was constantly being underclocked when using the schedutil governor.

**Reply by dormando** [reply]
Yes same answer. There's nothing special we do so whether or not the scheduler keeps the CPU engaged is luck of the draw. I only use performance governor when testing performance.
